src/Microsoft.Extensions.Primitives/StringValues.cs

```diff
@@ -18,17 +19,55 @@ public struct StringValues : IList<string>, IReadOnlyList<string>, IEquatable<St
 
         private readonly string _value;
         private readonly string[] _values;
-
+        private readonly Encoding _encoding;
+        private readonly byte[] _bytes;
+        
         public StringValues(string value)
         {
             _value = value;
             _values = null;
+            _encoding = null;
+            _bytes = null;
         }
 
         public StringValues(string[] values)
         {
             _value = null;
             _values = values;
+            _encoding = null;
+            _bytes = null;
+        }
+
+        private StringValues(string value, Encoding encoding)
+        {
+            _value = value;
+            _values = null;
+            _encoding = encoding;
+            _bytes = _encoding.GetBytes(value);
+        }
+
+        /// <summary>
+        /// Creates a new instance of <see cref="StringValues"/> with one string and compute its encoded bytes with the specified Encoding.
+        /// </summary>
+        /// <param name="value">The string to be encoded.</param>
+        /// <param name="encoding">The <see cref="Encoding"/> to be use when computing pre-encoded bytes.</param>
+        /// <returns>A <see cref="StringValues"/> which contains the pre-encoded bytes.</returns>
+        public static StringValues CreatePreEncoded(string value, Encoding encoding)
+        {
+            return new StringValues(value, encoding);
+        }
+
+        /// <summary>
+        /// Try to retrieve the pre-encoded bytes of the <see cref="StringValues"/>. The return value indicates whether pre-encoded bytes exist.
+        /// </summary>
+        /// <param name="bytes">If successful, <paramref name="bytes"/> contains the pre-encoded bytes of the <see cref="StringValues"/>.</param>
+        /// <param name="encoding">If successful, <paramref name="encoding"/> contains the <see cref="Encoding"/> used to compute the pre-encoded bytes.</param>
+        /// <returns><c>true</c> if <see cref="StringValues"/> contains pre-encoded bytes, otherwise <c>false</c>.</returns>
+        public bool TryGetPreEncoded(out byte[] bytes, out Encoding encoding)
```

Comment:
Couldn't we make `EncodedBytes` a public property and add a remark that says its only available for pre-encoded values? Two `out`s feels like bad design.